src/github.com/couchbase/sync_gateway/base/bucket_gocb.go

```diff
@@ -93,91 +97,416 @@ func (bucket CouchbaseBucketGoCB) GetName() string {
 	return bucket.spec.BucketName
 }
 
+func (bucket CouchbaseBucketGoCB) GetRaw(k string) (rv []byte, cas uint64, err error) {
+
+	var returnVal interface{}
+	cas, err = bucket.Get(k, &returnVal)
+	if returnVal == nil {
+		return nil, cas, err
+	}
+	return returnVal.([]byte), cas, err
+
+}
+
 func (bucket CouchbaseBucketGoCB) Get(k string, rv interface{}) (cas uint64, err error) {
+
 	bucket.singleOps <- struct{}{}
 	gocbExpvars.Add("SingleOps", 1)
 	defer func() {
 		<-bucket.singleOps
 		gocbExpvars.Add("SingleOps", -1)
 	}()
-	gocbExpvars.Add("Get", 1)
-	casGoCB, err := bucket.Bucket.Get(k, rv)
-	return uint64(casGoCB), err
+	worker := func() (shouldRetry bool, err error, value interface{}) {
+
+		gocbExpvars.Add("Get", 1)
+		casGoCB, err := bucket.Bucket.Get(k, rv)
+		shouldRetry = shouldRetryGoCBOp(err)
+
+		return shouldRetry, err, uint64(casGoCB)
+
+	}
+
+	sleeper := CreateDoublingSleeperFunc(
+		bucket.spec.MaxNumRetries,
+		bucket.spec.InitialRetrySleepTimeMS,
+	)
+
+	// Kick off retry loop
+	description := fmt.Sprintf("Get %v", k)
+	err, result := RetryLoop(description, worker, sleeper)
+
+	// If the retry loop returned a nil result, set to 0 to prevent type assertion on nil error
+	if result == nil {
+		result = uint64(0)
+	}
+
+	// Type assertion of result
+	cas, ok := result.(uint64)
+	if !ok {
+		LogPanic("Error doing type assertion of %v into a uint64", result)
+	}
+
+	return cas, err
+
 }
 
-func (bucket CouchbaseBucketGoCB) GetRaw(k string) (rv []byte, cas uint64, err error) {
+// Retry up to the retry limit, then return.  Does not retry items if they had CAS failures,
+// and it's up to the caller to handle those.
+func (bucket CouchbaseBucketGoCB) SetBulk(entries []*sgbucket.BulkSetEntry) (err error) {
 
-	bucket.singleOps <- struct{}{}
-	gocbExpvars.Add("SingleOps", 1)
-	defer func() {
-		<-bucket.singleOps
-		gocbExpvars.Add("SingleOps", -1)
-	}()
+	// Create the RetryWorker for BulkSet op
+	worker := bucket.newSetBulkRetryWorker(entries)
 
-	var returnVal interface{}
+	// this is the function that will be called back by the RetryLoop to determine
+	// how long to sleep before retrying (uses backoff)
+	sleeper := CreateDoublingSleeperFunc(
+		bucket.spec.MaxNumRetries,
+		bucket.spec.InitialRetrySleepTimeMS,
+	)
+
+	// Kick off retry loop
+	description := fmt.Sprintf("SetBulk with %v entries", len(entries))
+	err, _ = RetryLoop(description, worker, sleeper)
+
+	return err
+
+}
+
+func (bucket CouchbaseBucketGoCB) newSetBulkRetryWorker(entries []*sgbucket.BulkSetEntry) RetryWorker {
+
+	pendingKeys := createPendingKeysFromEntries(entries)
+	entriesKeysToIndex := createEntriesKeyToIndex(entries)
+
```

Comment:
I'm wondering whether we could just use entries as-is, instead of splitting into a keySet and an index lookup array.  Any reason that pendingKeysBatches and processEntriesBatch couldn't just work directly against entries?